gen/spec.go

```diff
@@ -199,6 +224,176 @@ type traversal interface {
+		)
+		t.Fuse(data)
+
+		next(t, f.FieldElem)
+	}
+}
+
+func genStructFieldsParser(t traversalAssigner, p printer, fields []StructField) {
+	const (
+		fieldBytes = "field_b"
+		fieldInt   = "field_i"
+		fieldUint  = "field_u"
+		typ        = "typ"
+	)
+
+	groups := groupFieldsByType(fields)
+	hasUint := len(groups[msgp.UintType]) > 0
+	hasInt := len(groups[msgp.IntType]) > 0
+	hasStr := len(groups[msgp.StrType]) > 0
+	singleType := len(groups) == 1
+
+	if hasStr {
+		t.declareOnce(p, fieldBytes, "[]byte")
+	}
+	if hasUint {
+		t.declareOnce(p, fieldUint, "uint64")
+
+		// Append to int fields also uint fields that do not overflow int64.
+		// This is necessary because uint field with value <= (1<<7)-1 could be serialized as fixint.
+		// and become a msgp.IntType. This is done for best compatibility with other libraries
+		// (and with other languages). That is, some endpoint could serialize uint16 key with value <= (1<<7)-1 as
+		// real msgpack uint16, but also could serialize it like fixint.
+		for _, f := range groups[msgp.UintType] {
+			v := f.FieldTag.(uint64)
+			if v <= math.MaxInt64 {
+				groups[msgp.IntType] = append(groups[msgp.IntType], f)
+				hasInt = true
+			}
+		}
+	}
+	if hasInt {
+		t.declareOnce(p, fieldInt, "int64")
+	}
+	if !singleType || hasUint {
+		t.declareOnce(p, typ, "msgp.Type")
+	}
+
+	sz := randIdent()
+	p.declare(sz, u32)
+	t.assignAndCheck(sz, mapHeader)
+	p.printf("\nfor %s > 0 {\n%s--", sz, sz)
+	switch {
+	case singleType && hasStr:
+		t.assignAndCheck(fieldBytes, mapKey)
+		switchFieldKeysStr(t, p, fields, fieldBytes)
+	case singleType && hasInt:
+		t.assignAndCheck(fieldInt, "Int64")
+		switchFieldKeys(t, p, fields, fieldInt)
+	default:
+		// switch on inferred type of next field
+		t.nextTypeAndCheck(typ)
+		p.printf("\nswitch %s {", typ)
+		if hasUint {
+			p.print("\ncase msgp.UintType:")
+			t.assignAndCheck(fieldUint, "Uint64")
+			switchFieldKeys(t, p, groups[msgp.UintType], fieldUint)
+		}
+		if hasInt {
+			p.print("\ncase msgp.IntType:")
+			t.assignAndCheck(fieldInt, "Int64")
+			switchFieldKeys(t, p, groups[msgp.IntType], fieldInt)
+		}
+		if hasStr {
+			// double case is done for backward compatibility with previous implementation
+			p.print("\ncase msgp.StrType, msgp.BinType:")
+			t.assignAndCheck(fieldBytes, mapKey)
+			switchFieldKeysStr(t, p, groups[msgp.StrType], fieldBytes)
+		}
+		p.print("\ndefault:")
+		t.skipAndCheck()
+		p.closeblock() // close switch
+	}
+	p.closeblock() // close loop
+}
+
+func groupFieldsByType(fields []StructField) map[msgp.Type][]StructField {
+	groups := make(map[msgp.Type][]StructField, len(fields))
+	for _, f := range fields {
+		var t msgp.Type
+		switch f.FieldTag.(type) {
+		case int, int8, int16, int32, int64:
+			t = msgp.IntType
+		case uint, uint8, uint16, uint32, uint64:
+			t = msgp.UintType
+		case string:
+			t = msgp.StrType
+		default:
+			panic(fmt.Errorf("field %q has unknown tag type %T", f.FieldName, f.FieldTag))
```

Comment:
There are already few cases when generator panics. For example, [here](https://github.com/tinylib/msgp/blob/master/gen/spec.go#L219). So is it really necessary to print error and do some of `os.Exit(1)`?